clients/da-vinci-client/src/main/java/com/linkedin/davinci/validation/KafkaDataIntegrityValidator.java

```diff
@@ -97,4 +122,37 @@ public void checkMissingMessage(
     producerTracker.checkMissingMessage(consumerRecord, errorMetricCallback, this.kafkaLogCompactionDelayInMs);
   }
 
+  void clearExpiredState() {
+    if (this.maxAgeInMs != DISABLED) {
+      long currentTime = time.getMilliseconds();
+      long maxRecordProducerTimestamp = currentTime - this.maxAgeInMs;
+      if (this.oldestRetainedProducerTS < maxRecordProducerTimestamp) {
+        synchronized (this) {
+          if (this.oldestRetainedProducerTS < maxRecordProducerTimestamp) {
+            int numberOfClearedGUIDs = 0;
+            long newOldestRetainedProducerTS = Long.MAX_VALUE;
+            Iterator<ProducerTracker> iterator = this.producerTrackerMap.values().iterator();
+            ProducerTracker producerTracker;
+            while (iterator.hasNext()) {
+              producerTracker = iterator.next();
+              long currentProducerOldestRetainedTS = producerTracker.clearExpiredState(maxRecordProducerTimestamp);
+              if (currentProducerOldestRetainedTS == Long.MAX_VALUE) {
+                iterator.remove();
+                numberOfClearedGUIDs++;
+              } else {
+                newOldestRetainedProducerTS = Math.min(newOldestRetainedProducerTS, currentProducerOldestRetainedTS);
+              }
+            }
+            if (newOldestRetainedProducerTS == Long.MAX_VALUE) {
+              /** We only update the {@link #oldestRetainedProducerTS} if there is any state we retained. */
+              this.oldestRetainedProducerTS = newOldestRetainedProducerTS;
```

Comment:
How does the state of oldestRetainedProducerTS get reset to something less then Long.MAX_VALUE after this state?  It seems like if this happens, then this function won't run again ever.

Reply:
Oh that's not right... it's meant to be if !=... I'll try to extend the unit test to demonstrate that this is broken...

Reply:
The latest diff has the test and the fix. Thanks for catching this bug!